RFC: IdDict replaces ObjectIdDict

mauro3 opened the pull request:
Third time lucky?  This is a the third attempt, after #25196 and #24932.  After letting Jeff's comment in #25196 sink in, I decided to not wrap ObjectIdDict but to directly furbish its getters and setters with type assertions.  This is cleaner than either of my previous attempts.  However, this doesn't quite work yet:
the tests run locally but then the julia process does not exit (CPU is back to 0%).  I suspect it has something to do with the C-part.

Also, I removed a lot of the nonspecialize stuff. Not sure what the impact of that is.

mauro3 committed:
Renamed ObjectIdDict -> IdDict{K,V} and added {K,V}

- also updated constuctors to be in-line with Dict
- added some extra tests

mauro3 commented on base/abstractdict.jl:
```diff
@@ -502,108 +502,149 @@ push!(t::AbstractDict, p::Pair) = setindex!(t, p.second, p.first)
         d
     end
+end
+const _IdDict = IdDict{Any,Any} # this is needed to make src/dump.c and src/staticdata.c work
```
This is iffy.  I did this to make it compile, see comment below.

mauro3 commented on src/dump.c:
```diff
@@ -2297,7 +2297,7 @@ JL_DLLEXPORT int jl_save_incremental(const char *fname, jl_array_t *worklist)
     ptrhash_put(&backref_table, jl_main_module, (char*)HT_NOTFOUND + 1);
     backref_table_numel = 1;
-    jl_idtable_type = jl_base_module ? jl_get_global(jl_base_module, jl_symbol("ObjectIdDict")) : NULL;
+    jl_idtable_type = jl_base_module ? jl_get_global(jl_base_module, jl_symbol("_IdDict")) : NULL;
```
This then uses https://github.com/JuliaLang/julia/pull/25210/files#r158050304

mauro3 commented on src/staticdata.c:
```diff
@@ -1205,7 +1205,7 @@ static void jl_save_system_image_to_stream(ios_t *f)
     }
 
-    jl_idtable_type = jl_base_module ? jl_get_global(jl_base_module, jl_symbol("ObjectIdDict")) : NULL;
+    jl_idtable_type = jl_base_module ? jl_get_global(jl_base_module, jl_symbol("_IdDict")) : NULL;
```
This then uses https://github.com/JuliaLang/julia/pull/25210/files#r158050304

JeffBezanson replied:
The rehashing will now need to be done for all instantiations of `IdDict`, not just `IdDict{Any,Any}`.

mauro3 replied:
But just replacing it with `jl_symbol("IdDict")` does not work.  Julia compiles fine but the parallel stuff does not work (and thus the test don't run either). On eac2ff7:
```
julia> using Distributed; addprocs(2)
      From worker 2:    Master process (id 1) could not connect within 60.0 seconds.Worker 2 terminated.
ERROR: 
      From worker 2:    exiting.StackOverflowError:
haskey at ./abstractdict.jl:553 [inlined]
is_root_module at ./loading.jl:371 [inlined]
serialize_mod_names(::Distributed.ClusterSerializer{TCPSocket}, ::Module) at ./serialize.jl:355
serialize_mod_names(::Distributed.ClusterSerializer{TCPSocket}, ::Module) at ./serialize.jl:358 (repeats 97 times)

...and 1 more exception(s).


Stacktrace:ERROR (unhandled task failure): Version read failed. Connection closed by peer.

 [1] sync_end() at ./task.jl:300
 [2] macro expansion at ./task.jl:316 [inlined]
 [3] #addprocs_locked#40(::NamedTuple{(),Tuple{}}, ::Function, ::Distributed.LocalManager) at /home/mauro/julia/julia-master/usr/share/julia/site/v0.7/Distributed/src/cluster.jl:401
 [4] addprocs_locked at /home/mauro/julia/julia-master/usr/share/julia/site/v0.7/Distributed/src/cluster.jl:372 [inlined]
 [5] #addprocs#39(::NamedTuple{(),Tuple{}}, ::Function, ::Distributed.LocalManager) at /home/mauro/julia/julia-master/usr/share/julia/site/v0.7/Distributed/src/cluster.jl:365
 [6] addprocs at /home/mauro/julia/julia-master/usr/share/julia/site/v0.7/Distributed/src/cluster.jl:361 [inlined]
 [7] #addprocs#268 at /home/mauro/julia/julia-master/usr/share/julia/site/v0.7/Distributed/src/managers.jl:316 [inlined]
 [8] addprocs(::Int64) at /home/mauro/julia/julia-master/usr/share/julia/site/v0.7/Distributed/src/managers.jl:315
 [9] top-level scope
```

I think the problem is the re-hashing:
```
julia> haskey(Base.module_keys, Main) # true about half the time
false

julia> Base.rehash!(Base.module_keys);

julia> haskey(Base.module_keys, Main)
true
```

JeffBezanson commented on base/abstractdict.jl:
```diff
@@ -502,108 +502,149 @@ push!(t::AbstractDict, p::Pair) = setindex!(t, p.second, p.first)
+IdDict() = IdDict{Any,Any}()
+IdDict(kv::Tuple{}) = IdDict()
+
+IdDict(ps::Pair{K,V}...)           where {K,V} = IdDict{K,V}(ps)
```
We probably don't need all this; we can just default to `Any,Any`.

mauro3 replied:
That will make IdDict construction different from Dict construction. Is that desired?

JeffBezanson replied:
It already is; previously all `ObjectIdDicts` were `AbstractDict{Any,Any}`s. ObjectIdDicts are also often heterogeneous, and I don't see a need to break those use cases (i.e. where different-typed keys are added to a dict after construction).

mauro3 replied:
I think it would make Julia more consistent (c.f. #20402) to have IdDict constructors behave like Dict constructors (and also like Arrays, and presumably many others which try to guess a tight type parameter according to the arguments).  Also note that of the uses in Base, only one would need updating to`IdDict{Any,Any}(...)`.

That would also make a possible future transition to a more unified dict-backend along the lines of #24932 easier.

JeffBezanson commented on base/abstractdict.jl:
```diff
@@ -502,108 +502,149 @@ push!(t::AbstractDict, p::Pair) = setindex!(t, p.second, p.first)
 
-get(t::ObjectIdDict, @nospecialize(key), @nospecialize(default)) =
-    ccall(:jl_eqtable_get, Any, (Any, Any, Any), t.ht, key, default)
+function get(d::IdDict{K,V}, key::K, default) where {K, V}
```
The key doesn't need to be of type `K`, since it's ok for it not to be found here. The `@nospecialize` annotations can also probably be retained.

mauro3 replied:
Ok, so you'd prefer a KeyError to a MethodError.  Would it then also make sense to throw a KeyError for `setindex!`?

Also, should I retain all `@nospecialize` or just the one you mentioned here?

JeffBezanson replied:
Maybe this is similar enough to the `ArgumentError("$key0 is not a valid key for type $K")` that `Dict` throws that we should use the same error here?

And yes, we should keep all the `@nospecialize`. The goal of this should just be to allow type parameters in `ObjectIdDict` with the minimum changes.

JeffBezanson commented on base/abstractdict.jl:
```diff
@@ -502,108 +502,149 @@ push!(t::AbstractDict, p::Pair) = setindex!(t, p.second, p.first)
+    v = ccall(:jl_eqtable_get, Any, (Any, Any, Any), d.ht, key, default)
+    v===default ? default : v::V
+end
+function getindex(d::IdDict{K,V}, key::K) where {K, V}
```
Similarly, there should be a `KeyError` here if the key is not of type `K`.

JeffBezanson commented on base/abstractdict.jl:
```diff
@@ -502,108 +502,149 @@ push!(t::AbstractDict, p::Pair) = setindex!(t, p.second, p.first)
     # TODO: this can underestimate `ndel`
-    val === default || (t.ndel += 1)
-    return val
+    if val==default
```
`===` ?

JeffBezanson commented:
:+1: Yes I think this is the right approach.

mauro3 commented:
Thanks for the comments.  However, I'm stuck on https://github.com/JuliaLang/julia/pull/25210#discussion_r158142358.

andyferris commented on base/abstractdict.jl:
```diff
@@ -502,108 +502,140 @@ push!(t::AbstractDict, p::Pair) = setindex!(t, p.second, p.first)
+    IdDict{K,V}(d::IdDict{K,V}) where {K, V} = new{K,V}(copy(d.ht))
+    function IdDict{K,V}(d::IdDict{K}) where {K, V}
+        d = IdDict{K,V}()
+        sizehint!(d, length(pairs))
```
This method seems wrong. Is it a mistaken copy-paste of the second method above?

mauro3 replied:
Yes, thanks, the `pairs` is wrong.  But the method itself is needed:

My thinking here was to provide a method to allow something like:
```
di = IdDict{Symbol,Int}(:A=>1)
IdDict{Symbol,Float64}(di)
```
the loop doing the `setindex!` will call convert on the integer values of `di`.

I'll fix it and add a test.

mauro3 replied:
Actually, it is not needed after all as it will call into the general `function IdDict{K,V}(itr) ...` method.  Thanks!

mauro3 commented:
I can spend some more time on this tomorrow to try and hunt down the right invocation in [dump.c](https://github.com/JuliaLang/julia/pull/25210/files#diff-4be7427a88e9b4e136acb0f3385267d0R1208) and staticdata.c to make automatic re-hashing work again.  See https://github.com/JuliaLang/julia/pull/25210#discussion_r158142358 for details. I suspect that the problem is that `IdDict{K,V}` is not a symbol anymore, so `jl_symbol("IdDict")` doesn't catch it (nor `jl_symbol("IdDict{K,V}")`).

However, I have no clue on how to make this work so it will be blind trial and error.  If someone finds the time to point me in the right direction, that would be much appreciated.

mauro3 commented:
After an educational morning digging through the C-part of Julia, I found a way to make it work 🎆.  The equality test needed [here](https://github.com/JuliaLang/julia/blob/6c94bd84a275b96e4f7d8d77d1563561978b686c/src/staticdata.c#L738) (and also in dump.c) had to be converted into a subtype test.  I did not find a function for this so I took inspiration from, e.g., array subtyping tests [here](https://github.com/JuliaLang/julia/blob/6c94bd84a275b96e4f7d8d77d1563561978b686c/src/julia.h#L957).  The test suite now passes without a hickup (barring any rebase hickups before the push), but definitely needs checking by someone who groks this.

Other things I did:
- I added all `@nospecialize` annotations back in (and added them `get!` also, as it seems inconsistent that it should be lacking them).
- I added the normal dict-like constructors back in as they make Julia more consistent (IMO), [here](https://github.com/JuliaLang/julia/pull/25210#discussion_r158384344) my argument.  But I can just remove the last two commits, to revert to good old ObjectIdDict-days.

mauro3 commented:
I'll be away until 3 Jan.  In the unlikely event that this acquires urgency, feel free modify this PR.

mauro3 commented:
This still needs a review, in particular the C-part.

JeffBezanson commented on base/abstractdict.jl:
```diff
@@ -514,107 +514,144 @@ function convert(::Type{T}, x::AbstractDict) where T<:AbstractDict
-    pop!(t, key, secret_table_token)
-    t
+function delete!(d::IdDict{K}, @nospecialize(key)) where K
+    !isa(key, K) && throw(KeyError(key))
```
I don't think any other methods of `delete!` throw KeyErrors.

JeffBezanson commented:
This looks great, let's do it!

mauro3 commented:
Thanks Jeff. I removed the KeyError from `delete!`, rebased and squashed this into one commit.  I also added two NEWS.md entries.

So this is ready to be merged when CI passes, unless it is deemed necessary to run nanosoldier on this, although I think there are no direct ObjectIdDict tests in there. I have re-run some primitive performance tests in https://gist.github.com/mauro3/ada59b1fd9a09202631edccf255c4e1b#gistcomment-2323560, and find no regression.

mauro3 commented:
Looks like the one failure on AppVeyor was a time-out (finished at 2h exactly).

JeffBezanson commented:
Yes that's happening on basically every build lately.

StefanKarpinski commented:
The doc builds take so long these days that AppVeyor times out all the time. It's a pretty urgent problem at the moment. cc @mortenpi: it would be great to get a fix – any fix for this merged asap.

JeffBezanson commented:
Thinking about it some more, and while updating some code to use this, I wonder if all `IdDict(...)` constructors should make an `IdDict{Any,Any}` (instead of deriving type parameters from their argument types like `Dict`). The advantage is that people can use `IdDict` as a drop-in replacement for `ObjectIdDict` in more cases, and we avoid extra specialization that's likely not too valuable (since there's no specialization on storage).

StefanKarpinski commented:
The consistency argument is pretty compelling, especially given that initializing an `IdDict` with values to specialize on at all is a bit strange – and if you do, perhaps you want the tighter type.

JeffBezanson commented:
Fair enough; it's ok to keep it as-is.
